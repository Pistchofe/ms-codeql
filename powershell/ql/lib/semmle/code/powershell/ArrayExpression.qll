import powershell

class ArrayExpr extends @array_expression, Expr {
  override SourceLocation getLocation() { array_expression_location(this, result) }

  StmtBlock getStmtBlock() { array_expression(this, result) }

  /**
   * Gets the i'th element of this `ArrayExpr`, if this can be determined statically.
   *
   * See `getStmtBlock` when the array elements are not known statically.
   */
  Expr getElement(int i) {
    result =
      unique( | | this.getStmtBlock().getAStmt()).(CmdExpr).getExpr().(ArrayLiteral).getElement(i)
  }

  /**
   * Gets an element of this `ArrayExpr`, if this can be determined statically.
   *
   * See `getStmtBlock` when the array elements are not known statically.
   */
  Expr getAnElement() { result = this.getElement(_) }

  override string toString() { result = "@(...)" }
}
