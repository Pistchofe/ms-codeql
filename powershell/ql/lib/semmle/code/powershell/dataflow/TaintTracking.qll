/**
 * Provides classes for performing local (intra-procedural) and
 * global (inter-procedural) taint-tracking analyses.
 */
module TaintTracking {
  import semmle.code.powershell.dataflow.internal.TaintTrackingImpl::Public
  private import semmle.code.powershell.dataflow.internal.DataFlowImplSpecific
  private import semmle.code.powershell.dataflow.internal.TaintTrackingImplSpecific
  private import codeql.dataflow.TaintTracking
  private import powershell
  import TaintFlowMake<Location, PowershellDataFlow, PowershellTaintTracking>
}
