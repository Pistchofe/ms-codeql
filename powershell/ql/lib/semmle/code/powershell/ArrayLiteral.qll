import powershell

class ArrayLiteral extends @array_literal, Expr {
  override SourceLocation getLocation() { array_literal_location(this, result) }

  Expr getElement(int index) { array_literal_element(this, index, result) }

  Expr getAnElement() { array_literal_element(this, _, result) }

  override string toString() { result = "...,..." }
}
