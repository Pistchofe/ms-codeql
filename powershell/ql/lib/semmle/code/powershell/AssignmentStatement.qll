import powershell

class AssignmentStatement extends @assignment_statement instanceof Statement
{
    SourceLocation getLocation()
    {
        assignment_statement_location(this, result)
    }

    int getKind()
    {
        assignment_statement(this, result, _, _)
    }

    Expression getLeftHandSide()
    {
        assignment_statement(this, _, result, _)
    }

    Statement getRightHandSide()
    {
        assignment_statement(this, _, _, result)
    }

    string toString(){
        result = "AssignmentStatement at: " + this.getLocation().toString()
    }
}