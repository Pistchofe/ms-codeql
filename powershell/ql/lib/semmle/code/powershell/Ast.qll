import powershell

class Ast extends @ast {
  string toString() { none() }

  Ast getParent() { parent(this, result) }

  Location getLocation() { none() }
}
