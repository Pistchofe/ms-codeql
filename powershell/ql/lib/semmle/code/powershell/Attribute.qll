import powershell

class Attribute extends @attribute, AttributeBase {
  override string toString() { result = this.getName() }

  override SourceLocation getLocation() { attribute_location(this, result) }

  string getName() { attribute(this, result, _, _) }

  int getNumNamedArguments() { attribute(this, _, result, _) }

  int getNumPositionalArguments() { attribute(this, _, _, result) }

  NamedAttributeArgument getNamedArgument(int i) { attribute_named_argument(this, i, result) }

  NamedAttributeArgument getANamedArgument() { result = this.getNamedArgument(_) }

  int getNumberOfArguments() { result = count(this.getAPositionalArgument()) }

  Expr getPositionalArgument(int i) { attribute_positional_argument(this, i, result) }

  Expr getAPositionalArgument() { result = this.getPositionalArgument(_) }

  int getNumberOfPositionalArguments() { result = count(this.getAPositionalArgument()) }
}
