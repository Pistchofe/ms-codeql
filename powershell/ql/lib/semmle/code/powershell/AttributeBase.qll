import powershell

class AttributeBase extends @attribute_base, Ast { }
