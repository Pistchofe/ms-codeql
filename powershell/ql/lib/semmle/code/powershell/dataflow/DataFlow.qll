/**
 * Provides classes for performing local (intra-procedural) and
 * global (inter-procedural) data flow analyses.
 */

import powershell

module DataFlow {
  private import internal.DataFlowImplSpecific
  private import codeql.dataflow.DataFlow
  import DataFlowMake<Location, PowershellDataFlow>
  import internal.DataFlowImpl
}
