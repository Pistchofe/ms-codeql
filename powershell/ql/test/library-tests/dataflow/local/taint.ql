import powershell
import semmle.code.powershell.dataflow.TaintTracking
import semmle.code.powershell.dataflow.DataFlow

from DataFlow::Node pred, DataFlow::Node succ
where TaintTracking::localTaintStep(pred, succ)
select pred, succ