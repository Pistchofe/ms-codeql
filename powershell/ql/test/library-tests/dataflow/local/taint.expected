| test.ps1:1:1:1:4 | a1 | test.ps1:2:6:2:9 | a1 |
| test.ps1:1:7:1:13 | Source | test.ps1:1:1:1:4 | a1 |
| test.ps1:1:7:1:13 | Source | test.ps1:1:1:1:13 | ...=... |
| test.ps1:2:1:2:9 | Sink | test.ps1:2:1:2:9 | pre-return value for Sink |
| test.ps1:2:1:2:9 | Sink | test.ps1:2:1:2:9 | pre-return value for Sink |
| test.ps1:2:1:2:9 | implicit unwrapping of Sink | test.ps1:1:1:17:8 | return value for test.ps1 |
| test.ps1:2:1:2:9 | pre-return value for Sink | test.ps1:2:1:2:9 | implicit unwrapping of Sink |
| test.ps1:2:1:2:9 | pre-return value for Sink | test.ps1:2:1:2:9 | implicit unwrapping of Sink |
| test.ps1:4:1:4:3 | b | test.ps1:5:4:5:6 | b |
| test.ps1:4:6:4:13 | GetBool | test.ps1:4:1:4:3 | b |
| test.ps1:4:6:4:13 | GetBool | test.ps1:4:1:4:13 | ...=... |
| test.ps1:5:4:5:6 | b | test.ps1:10:14:10:16 | b |
| test.ps1:6:5:6:8 | a2 | test.ps1:8:6:8:9 | a2 |
| test.ps1:6:11:6:17 | Source | test.ps1:6:5:6:8 | a2 |
| test.ps1:6:11:6:17 | Source | test.ps1:6:5:6:17 | ...=... |
| test.ps1:8:1:8:9 | Sink | test.ps1:8:1:8:9 | pre-return value for Sink |
| test.ps1:8:1:8:9 | Sink | test.ps1:8:1:8:9 | pre-return value for Sink |
| test.ps1:8:1:8:9 | implicit unwrapping of Sink | test.ps1:1:1:17:8 | return value for test.ps1 |
| test.ps1:8:1:8:9 | pre-return value for Sink | test.ps1:8:1:8:9 | implicit unwrapping of Sink |
| test.ps1:8:1:8:9 | pre-return value for Sink | test.ps1:8:1:8:9 | implicit unwrapping of Sink |
| test.ps1:10:1:10:3 | c | test.ps1:11:6:11:8 | c |
| test.ps1:10:6:10:16 | [...]... | test.ps1:10:1:10:3 | c |
| test.ps1:10:6:10:16 | [...]... | test.ps1:10:1:10:16 | ...=... |
| test.ps1:10:6:10:16 | [...]... | test.ps1:10:6:10:16 | [...]... |
| test.ps1:10:14:10:16 | b | test.ps1:10:6:10:16 | [...]... |
| test.ps1:11:1:11:8 | Sink | test.ps1:11:1:11:8 | pre-return value for Sink |
| test.ps1:11:1:11:8 | Sink | test.ps1:11:1:11:8 | pre-return value for Sink |
| test.ps1:11:1:11:8 | implicit unwrapping of Sink | test.ps1:1:1:17:8 | return value for test.ps1 |
| test.ps1:11:1:11:8 | pre-return value for Sink | test.ps1:11:1:11:8 | implicit unwrapping of Sink |
| test.ps1:11:1:11:8 | pre-return value for Sink | test.ps1:11:1:11:8 | implicit unwrapping of Sink |
| test.ps1:11:6:11:8 | [post] c | test.ps1:13:7:13:9 | c |
| test.ps1:11:6:11:8 | c | test.ps1:13:7:13:9 | c |
| test.ps1:13:1:13:3 | d | test.ps1:14:6:14:8 | d |
| test.ps1:13:6:13:10 | (...) | test.ps1:13:1:13:3 | d |
| test.ps1:13:6:13:10 | (...) | test.ps1:13:1:13:10 | ...=... |
| test.ps1:13:6:13:10 | (...) | test.ps1:13:6:13:10 | (...) |
| test.ps1:13:7:13:9 | c | test.ps1:13:6:13:10 | (...) |
| test.ps1:13:7:13:9 | c | test.ps1:13:7:13:9 | c |
| test.ps1:14:1:14:8 | Sink | test.ps1:14:1:14:8 | pre-return value for Sink |
| test.ps1:14:1:14:8 | Sink | test.ps1:14:1:14:8 | pre-return value for Sink |
| test.ps1:14:1:14:8 | implicit unwrapping of Sink | test.ps1:1:1:17:8 | return value for test.ps1 |
| test.ps1:14:1:14:8 | pre-return value for Sink | test.ps1:14:1:14:8 | implicit unwrapping of Sink |
| test.ps1:14:1:14:8 | pre-return value for Sink | test.ps1:14:1:14:8 | implicit unwrapping of Sink |
| test.ps1:14:6:14:8 | [post] d | test.ps1:16:6:16:8 | d |
| test.ps1:14:6:14:8 | d | test.ps1:16:6:16:8 | d |
| test.ps1:16:1:16:3 | e | test.ps1:17:6:17:8 | e |
| test.ps1:16:6:16:12 | ...+... | test.ps1:16:1:16:3 | e |
| test.ps1:16:6:16:12 | ...+... | test.ps1:16:1:16:12 | ...=... |
| test.ps1:16:6:16:12 | ...+... | test.ps1:16:6:16:12 | ...+... |
| test.ps1:17:1:17:8 | Sink | test.ps1:17:1:17:8 | pre-return value for Sink |
| test.ps1:17:1:17:8 | Sink | test.ps1:17:1:17:8 | pre-return value for Sink |
| test.ps1:17:1:17:8 | implicit unwrapping of Sink | test.ps1:1:1:17:8 | return value for test.ps1 |
| test.ps1:17:1:17:8 | pre-return value for Sink | test.ps1:17:1:17:8 | implicit unwrapping of Sink |
| test.ps1:17:1:17:8 | pre-return value for Sink | test.ps1:17:1:17:8 | implicit unwrapping of Sink |
